addons/survey/controllers/main.py

```diff
@@ -728,39 +724,153 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
+            if not data:
+                break
+            model, row_id, ans_id = data.split(',')
+            row_id, ans_id = int(row_id), int(ans_id)
+            if model == 'A':
+                if ans_id in answer_cols_to_rows_mapping:
+                    answer_cols_to_rows_mapping[ans_id].append(row_id)
+                else:
+                    answer_cols_to_rows_mapping[ans_id] = [row_id]
+            elif model == 'L' and not row_id:
+                answer_user_input_line_ids.append(ans_id)
+
+        return answer_cols_to_rows_mapping, answer_user_input_line_ids
+
+    @staticmethod
+    def _get_value_with_operator(record, row_id=0):
+        """
+            Depending on the record question type, get the question value
+            and the domain operator to be applied for the filter matching
+
+            :param record: 'survey.question.answer' or 'survey.user_input.line' record
+            :return: value, operator
+        """
+        if record.question_id.question_type in ('simple_choice', 'multiple_choice', 'matrix'):
+            return f'{f"{row_id.value} : " if row_id else ""}{record.value}', None
+        return record[f'value_{record.question_id.question_type}'], {
+            'text_box': 'ilike',
+            'char_box': '=ilike',
+            'date': '=',
+            'datetime': '=',
+            'numerical_box': '=',
+        }[record.question_id.question_type]
+
+    @staticmethod
+    def _prepare_filters_data(record, row_id, value, model):
+        """
+            Prepare the common filters data between the 2 models
+            :param record: 'survey.question.answer' or 'survey.user_input.line' record
+        """
+        return {
+            'question_id': record.question_id.id,
+            'question': record.question_id.title,
+            'row_id': row_id,
+            'answer': value,
+            'model': model,
+            'answer_id': record.id,
+        }
+
+    @staticmethod
+    def _get_matching_user_input_lines(request, filters_domain_items, user_input_domain):
+        """
+            Retrieve the matching user input lines depending on the survey base domain and the filters domains
+        """
+        # Add filters domain to the base domain
+        if filters_domain_items:
+            all_required_lines_domains = [
+                [('user_input_line_ids', 'in', request.env['survey.user_input.line'].sudo()._search(filters_domain))]
+                for filters_domain in filters_domain_items
+            ]
+            user_input_domain = expression.AND([user_input_domain, *all_required_lines_domains])
+
+        # Get the matching user input lines
+        user_inputs_query = request.env['survey.user_input'].sudo()._search(user_input_domain)
+        user_input_lines = request.env['survey.user_input.line'].search([('user_input_id', 'in', user_inputs_query)])
+
+        return user_input_lines
+
+    @classmethod
+    def _extract_filters_data(cls, request, survey_id, post):
```

Comment:
To help readability, and (and understanding the diff a little easier too), this method should be above above all the utils it calls. (or all the utils should be below this higher level one).
If our route calls method A that calls method B that calls method C, that is the order they should be in the file too.